File: benchmarks/go/06_string_operations.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

func benchmarkStringConcat(n int) {
	parts := make([]string, n)
	for i := 0; i < n; i++ {
		parts[i] = strconv.Itoa(i)
	}
	start := time.Now()
	result := strings.Join(parts, ",")
	elapsed := time.Since(start).Seconds() * 1000
	fmt.Printf("  join %d strings: %.6fms, len=%d\n", n, elapsed, len(result))
}

func benchmarkStringSplit(n int) {
	parts := make([]string, n)
	for i := 0; i < n; i++ {
		parts[i] = strconv.Itoa(i)
	}
	bigStr := strings.Join(parts, ",")

	start := time.Now()
	tokens := strings.Split(bigStr, ",")
	elapsed := time.Since(start).Seconds() * 1000
	fmt.Printf("  split %d tokens: %.6fms, count=%d\n", n, elapsed, len(tokens))
}

func benchmarkStringReplace(n int) {
	base := strings.Repeat("hello world ", 1000)

	start := time.Now()
	var result string
	for i := 0; i < n; i++ {
		result = strings.ReplaceAll(base, "world", "tova")
	}
	_ = result
	elapsed := time.Since(start).Seconds() * 1000
	fmt.Printf("  replace x%d: %.6fms\n", n, elapsed)
}

func benchmarkStringSearch(n int) {
	haystack := strings.Repeat("abcdefghij", 10000)

	start := time.Now()
	foundCount := 0
	for i := 0; i < n; i++ {
		if strings.Contains(haystack, "fghij") {
			foundCount++
		}
	}
	elapsed := time.Since(start).Seconds() * 1000
	fmt.Printf("  contains x%d: %.6fms, found=%d\n", n, elapsed, foundCount)
}

func main() {
	fmt.Println("BENCHMARK: string_operations")
	benchmarkStringConcat(100000)
	benchmarkStringConcat(1000000)
	benchmarkStringSplit(100000)
	benchmarkStringSplit(1000000)
	benchmarkStringReplace(10000)
	benchmarkStringSearch(100000)
}

File: benchmarks/python/06_string_operations.py
import time

def benchmark_string_concat(n):
    parts = [str(i) for i in range(n)]
    start = time.perf_counter()
    result = ",".join(parts)
    elapsed = (time.perf_counter() - start) * 1000
    print(f"  join {n} strings: {elapsed}ms, len={len(result)}")

def benchmark_string_split(n):
    parts = [str(i) for i in range(n)]
    big_str = ",".join(parts)

    start = time.perf_counter()
    tokens = big_str.split(",")
    elapsed = (time.perf_counter() - start) * 1000
    print(f"  split {n} tokens: {elapsed}ms, count={len(tokens)}")

def benchmark_string_replace(n):
    base = "hello world " * 1000

    start = time.perf_counter()
    for _ in range(n):
        result = base.replace("world", "tova")
    elapsed = (time.perf_counter() - start) * 1000
    print(f"  replace x{n}: {elapsed}ms")

def benchmark_string_search(n):
    haystack = "abcdefghij" * 10000

    start = time.perf_counter()
    found_count = 0
    for _ in range(n):
        if "fghij" in haystack:
            found_count += 1
    elapsed = (time.perf_counter() - start) * 1000
    print(f"  contains x{n}: {elapsed}ms, found={found_count}")

print("BENCHMARK: string_operations")
benchmark_string_concat(100000)
benchmark_string_concat(1000000)
benchmark_string_split(100000)
benchmark_string_split(1000000)
benchmark_string_replace(10000)
benchmark_string_search(100000)
